lib/fileset.nix

```diff
@@ -0,0 +1,856 @@
+        seq start
+          (foldl' (prev: name:
+            let
+              subtree = tree.${name};
+
+              intermediate =
+                f prev "${indent}- ${name}${_prettyTreeSuffix subtree}";
+            in
+            if subtree == null then
+              # Don't print anything at all if this subtree is empty
+              prev
+            else if isAttrs subtree then
+              # A directory with explicit entries
+              # Do print this node, but also recurse
+              traceTreeAttrs intermediate "${indent}  " subtree
+            else
+              # Either a file, or a recursively included directory
+              # Do print this node but no further recursion needed
+              intermediate
+          ) start (attrNames tree));
+
+      intermediate =
+        if tree == null then
+          f start "${toString base} (empty)"
+        else
+          f start "${toString base}${_prettyTreeSuffix tree}";
+    in
+    if isAttrs tree then
+      traceTreeAttrs intermediate "" tree
+    else
+      intermediate;
+
+  # Coerce and normalise the bases of multiple file set values passed to user-facing functions
+  # Type: String -> [ { context :: String, value :: Any } ] -> { commonBase :: Path, trees :: [ <tree> ] }
+  _normaliseBase = function: list:
+    let
+      processed = map ({ context, value }:
+        let
+          fileset = _coerce function context value;
+        in {
+          inherit fileset context;
+          baseParts = deconstruct fileset._base;
+        }
+      ) list;
+
+      first = head processed;
+
+      commonComponents = foldl' (components: el:
+        if first.baseParts.root != el.baseParts.root then
+          throw "lib.fileset.${function}: Expected file sets to have the same filesystem root, but ${first.context} has root \"${toString first.baseParts.root}\" while ${el.context} has root \"${toString el.baseParts.root}\"."
+        else
+          commonPrefix components el.baseParts.components
+      ) first.baseParts.components (tail processed);
+
+      commonBase = construct {
+        root = first.baseParts.root;
+        components = commonComponents;
+      };
+
+      commonComponentsLength = length commonComponents;
+
+      trees = map (value:
+        _nestTree
+          commonBase
+          (drop commonComponentsLength value.baseParts.components)
+          value.fileset._tree
+      ) processed;
+    in
+    {
+      inherit commonBase trees;
+    };
+
+in {
+
+  /*
+  Import a file set into the store.
+  This function takes an attribute set as an argument with these attributes:
+
+  - `fileset`: The set of files to import into the store.
+    Use the `lib.fileset` combinator functions to define this value.
+
+  - `entryPoint`: The local directory that the resulting directory should focus on, meaning that if you `cd` into the resulting directory, all files in the fileset are accessible at paths relative to the `entryPoint`.
```

Comment:
Yeah this is hard to explain properly. I should reword this slightly, but the second part of this sentence is not too bad imo:

> meaning that if you `cd` into the resulting directory, all files in the fileset are accessible at paths relative to the `entryPoint`.

An example:

With a file set like this:
```
/home/user/my/project
- frontend
  - index.html
- common
  - vars.h
```

And `entryPoint` being `/home/user/my/project`, this works:
```
cd ${importToStore { ... }}
cat ./index.html ../common/vars.h
```

Underneath, a store path like this is created:
```
/nix/store/...
- frontend
  - index.html
- common
  - vars.h
```

However instead of pointing at it directly, the resulting string (via `outPath`) points at `/nix/store/.../frontend` instead, in order for the `cd` and `cat` to work correctly.